docs/developer/gvisor.md

```diff
@@ -0,0 +1,293 @@
+        },
+        "args": [
+            "/usr/bin/python3",
+            "-m",
+            "dangerzone.conversion.doc_to_pixels"
+        ],
+        "env": [
+            "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
+            "PYTHONPATH=/opt/dangerzone",
+            "TERM=xterm"
+        ],
+        "cwd": "/",
+        "capabilities": {
+            "bounding": [],
+            "effective": [
+                "CAP_SETUID",
+                "CAP_SETGID"
+            ],
+            "inheritable": [],
+            "permitted": [
+                "CAP_SETUID",
+                "CAP_SETGID"
+            ]
+        },
+        "rlimits": [
+            {
+                "type": "RLIMIT_NOFILE",
+                "hard": 4096,
+                "soft": 4096
+            }
+        ]
+    },
+    "root": {
+        "path": "rootfs",
+        "readonly": true
+    },
+    "hostname": "dangerzone",
+    "mounts": [
+        {
+            "destination": "/proc",
+            "type": "proc",
+            "source": "proc"
+        },
+        {
+            "destination": "/dev",
+            "type": "tmpfs",
+            "source": "tmpfs",
+            "options": [
+                "nosuid",
+                "noexec",
+                "nodev"
+            ]
+        },
+        {
+            "destination": "/sys",
+            "type": "tmpfs",
+            "source": "tmpfs",
+            "options": [
+                "nosuid",
+                "noexec",
+                "nodev",
+                "ro"
+            ]
+        },
+        {
+            "destination": "/tmp",
+            "type": "tmpfs",
+            "source": "tmpfs",
+            "options": [
+                "nosuid",
+                "noexec",
+                "nodev"
+            ]
+        },
+        {
+            "destination": "/home/dangerzone",
+            "type": "tmpfs",
+            "source": "tmpfs",
+            "options": [
+                "nosuid",
+                "noexec",
+                "nodev"
+            ]
+        },
+        {
+            "destination": "/usr/lib/libreoffice/share/extensions/",
+            "type": "tmpfs",
+            "source": "tmpfs",
+            "options": [
+                "nosuid",
+                "noexec",
+                "nodev"
+            ]
+        }
+    ],
+    "linux": {
+        "namespaces": [
+            {
+                "type": "pid"
+            },
+            {
+                "type": "network"
+            },
+            {
+                "type": "ipc"
+            },
+            {
+                "type": "uts"
+            },
+            {
+                "type": "mount"
+            }
+        ]
+    }
+}
+
+```
+
+## Security considerations
+
+* gVisor does not have an official release on Alpine Linux. The developers
+  provde gVisor binaries from a GCS bucket. In order to verify the integrity of
+  these binaries, they also provide a SHA-512 hash of the files.
```

Comment:
Like you, I'm a bit concerned that we are downloading the gVisor binaries directly from an external URL, which might be compromised and offer a nice attack vector when people are building their containers, especially as we are lowering the security of the outer container.

I'm not sure what type of practical solutions there are. Here one option I can think of:

The gVisor team is already already using [a GPG signing key](https://gvisor.dev/archive.key) for the apt repository, so we might use the signed binaries and check them against this key? 🤔

Reply:
Yeah, I'm afraid that this signing key is used to sign the release files for the APT repo. I'm afraid the actual binaries are not signed with this key.

Ultimately, we'll probably have to go down the same way we do for the PyPI packages, that is trust on first use (TOFU).

Reply:
Yeah, the Debian repo is a nice way forward. It so happens that we want to switch from the Alpine container image to the Debian image, so using the Debian repo will definitely fit in that case.

Reply:
> If you really really wanted, you could fetch the package from the apt repository using `apt download ...`, which verifies the signature, unpack the deb and extract the binary from it. I'm not sure if that provides any more meaningful trust though depending on how gvisor is signing their repository.

The gVisor release process pushes the same binary to both places. The `apt` solution is a bit better because the `runsc` binary is signed as part of the gVisor release process, as opposed to the just hash-based solution which could be susceptible to HTTPS MitM during the Dangerzone image build process.

Alternatively you can always build gVisor from source.

Reply:
> as opposed to the just hash-based solution which could be susceptible to HTTPS MitM (somehow) during the Dangerzone image build process.

Honestly, I'm less worried about an HTTPS attack, and more worried about a credential leak that would lead to a compromised GCS bucket. Especially given the fact that the checksum is served by the same bucket. That's why having a signature gives me more peace of mind.

For the last part (compiling gVisor from source), does the gVisor team offer signed tarballs, or signed Git tags, perhaps?

Reply:
@legoktm I see that Debian is lagging behind gVisor's official releases: https://tracker.debian.org/pkg/golang-gvisor-gvisor. I also see that a user has pointed them to the latest gVisor release: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1070301

Anw, this is just an fyi. It would be nice having a recent gVisor in the official repos.

Reply:
> For the last part (compiling gVisor from source), does the gVisor team offer signed tarballs, or signed Git tags, perhaps?

No, and even if it did, it would probably not be done in a way that meaningfully adds security (i.e. it would probably just be a bot blindly signing every commit, rather than individual contributors signing their own commits). The way the development workflow happens is just... too convoluted (to be generous) to make this work :(